test/pylib/pool.py

```diff
@@ -97,7 +101,7 @@ async def __aenter__(self):
                 return self.obj
 
             async def __aexit__(self, exc_type, exc, obj):
-                if self.obj:
+                if exc_type is not PoolDiscardItem and self.obj:
```

Comment:
I don't like the idea of using special exceptions for communicating with a context manager. In particular it prevents e.g. nested `try` blocks or nested uses of the context manager. If you want to communicate with the context manager, you can just add a method to the context manager and call it, i.e:
```python
async with (cm := self.suite.clusters.instance(logger)) as cluster:
     ...
     if cluster_broken:
          cm.discard()
     ...
```